### PR TITLE
Machine Layout

### PR DESCRIPTION
**Pull request by jacob-hughes**
This commit defines the structure of the virtual machine for the MIR interpreter. 
I think it's important for this to be provided on its own, as it
will make it easier to understand how the evaluation functions which step
through the MIR behave. (These will be provided incrementally in upcoming PRs.).
The code is heavily commented to explain why certain structures contain the
information they do.

The memory model is a rudimentary byte array, represented as a vector with each
item initialized to 0. The max size can be provided upon construction but I
haven't decided on what an appropriate value for this is yet.

One important aspect of the Machine's memory is that it is aligned according to
the min. required by the ABI. Allowing me to allocate ptrs guaranteed to be
correctly aligned.

    *See std::mem::align_of.

It is possible this may not be the correct alignment for my needs, the docs
state this function could return an alignment smaller than the one preferred
for the target architecture. I haven't thought too deeply about the implications
of this yet.

Memory cannot yet be deallocated, this will be implemented in the future.

### CODE REVIEW

**Comment by vext01** [comment]
```
One important aspect of the Machine's memory is that it is aligned according to
the min. required by the ABI.
```

Dumb question: Why is this important actually? Which operations in your interpreter require that certain bits of memory be aligned?

My guess is that to locate the start of a struct field, you might take the element's size and pad it up to the next natural alignment boundary. But is that strictly required? Perhaps it would be if you were expecting to pass the raw bytes inside your array to (e.g.) C code...

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#![feature(rustc_private, link_args)]
```
Comment (vext01) [inline src/lib.rs:1]: Don't forget to add licenses? The lawyers might come a knockin'!

File: src/machine.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+use rustc::mir::{Mir, Local};
+
+use interp::TypedVal;
+
+const X64_ALIGN_SIZE: u8 = 8;
+const X32_ALIGN_SIZE: u8 = 4;
+
+#[derive(Debug)]
+struct Frame<'tcx> {
+    // MIR is a CFG of a single function in rust, so it makes sense
```
Comment (vext01) [inline src/machine.rs]: Nitpick: `rust` -> `Rust`.

Comment (vext01) [reply src/machine.rs]: "Each MIR"?

File: src/machine.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    // for this to be in the frame.
+    mir: Mir<'tcx>,
+
+    // FIXME: Hacky... I can't find a way to actually access the private u32
+    // inside the `Local` tuple struct. `Local` does  implement the Hashable
+    // and PartialEQ trait however, so for  now it might just work to to use it
+    // as the key
```
Comment (vext01) [inline src/machine.rs:54]: Nitpick: Comments have full-stops, or not?

Comment (vext01) [reply src/machine.rs:54]: Yeah, `newtype_index!` is scary. Looks like the u32 is indeed private though.

File: src/machine.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            locals: HashMap::new(),
+            mir:    mir
+        }
+    }
+
+    fn set_local(&mut self, local: Local, val: Vec<u8>) {
+        self.locals.insert(local, val);
+    }
+
+    fn get_local(&self, local: Local) -> &Vec<u8> {
```
Comment (vext01) [inline src/machine.rs:70]: Do you plan on having error handling? If you do, probably one for another PR, but worth thinking about.

File: src/machine.rs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        self.locals.get(&local).unwrap()
+    }
+}
+
+// Represents a pointer into some kind of memory
+// TODO: Extend to work with static constructs
+enum Address {
```
Comment (vext01) [inline src/machine.rs:77]: Can you add comments onto the end of your enum variants?

I'm assuming the `usize` is an offset into the heap?

File: src/machine.rs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    // instantiation. The memory struct only cares about preserving alignment
+    // when handing out ptrs into it at runtime. Interpreting stored memory
+    // properly is the responsibility of those interfacing with it.
+    memory: Vec<u8>,
+
+    // Points to the last position in memory which is actually in use.
+    // Do not use directly to get a fresh ptr into memory.
+    // Instead use `next_aligned_ptr` method for a fresh correctly aligned ptr
+    // into memory for the target architecture.
+    next_free: usize,
```
Comment (vext01) [inline src/machine.rs:94]: Can memory fragment?

File: src/machine.rs [hunk 7]
```diff
@@ -0,0 +1,148 @@
+use std::collections::HashMap;
+use std::mem::align_of;
+
+use rustc::mir::{Mir, Local};
+
+use interp::TypedVal;
+
+const X64_ALIGN_SIZE: u8 = 8;
```
Comment (vext01) [inline src/machine.rs]: Unused constants?

File: src/machine.rs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            next_free: 0,
+            target_alignment: alignment
+        }
+    }
+
+    // Returns a pointer to next free block of memory of size n. Begins at index
+    // of the heap pointer.
+    //
+    // TODO: There is no free list so once a block of memory is allocated it is
+    // not yet possible to free it.
```
Comment (vext01) [inline src/machine.rs:113]: Ah, so memory can't yet fragment. OK.

File: src/machine.rs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    pub fn next_aligned_ptr(&mut self, size: usize) -> usize {
+        if (self.next_free + size) >= self.memory.len() {
```
Comment (vext01) [inline src/machine.rs]: This check needs to happen *after* alignment.

File: src/machine.rs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+            panic!("Out of memory.")
+        }
+
+        let r = self.next_free % self.target_alignment;
```
Comment (vext01) [inline src/machine.rs]: Is this alignment code right?

Suppose the next free space is 15 and we are aligning to 8.

 * 15 % 8 = 7
 * 15 + 7 = 22 (not aligned).

I suspect you want something like this:
https://play.rust-lang.org/?gist=4a1448d96b2491daf3a9163250a50d16&version=stable

Double check this, maths isn't my strong point!

**Review by vext01 — DISMISSED** [review]
Part 1 of my review.

File: src/lib.rs (hunk 1 above)
Comment (ltratt) [reply src/lib.rs:1]: I suggest copy-and-pasting the UPL license from one of our other projects. That said, you don't *need* to put a copyright license on things: you (well, the College in this case...) automatically owns the copyright. Software licenses are useful because they allow you to voluntarily give up some of the rights that copyright law gives you automatically.

File: src/machine.rs (hunk 3 above)
Comment (ltratt) [reply src/machine.rs:54]: We really need access to that number, because otherwise we're going to be stuck with a (very slow) `HashMap` rather than a (very fast) array. [Plus if you have numbers, there are other tricks you can do even before you think about arrays!]

File: src/machine.rs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    // inside the `Local` tuple struct. `Local` does  implement the Hashable
+    // and PartialEQ trait however, so for  now it might just work to to use it
+    // as the key
+    locals: HashMap<Local, Vec<u8>>
+}
+
+impl<'tcx> Frame<'tcx> {
+    fn new(mir: Mir<'tcx>) -> Frame {
+        Frame {
+            locals: HashMap::new(),
```
Comment (ltratt) [inline src/machine.rs:61]: Not that you need to do it in this PR, but you might want to lazily allocate this list and/or not allocate it if there are no local variables. Some small closures in tight loops won't have any locals, and that sort of saving can become useful.

File: src/machine.rs (hunk 4 above)
Comment (ltratt) [reply src/machine.rs:70]: Famous last words, but I don't think this function can fail. rustc will a) have checked that all the variables exist b) won't let you read a variable before it's written to.

File: src/machine.rs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+            mir:    mir
+        }
+    }
+
+    fn set_local(&mut self, local: Local, val: Vec<u8>) {
+        self.locals.insert(local, val);
+    }
+
+    fn get_local(&self, local: Local) -> &Vec<u8> {
+        self.locals.get(&local).unwrap()
```
Comment (ltratt) [inline src/machine.rs]: I think you can just do `self.locals[local]`?

File: src/machine.rs [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    // Instead use `next_aligned_ptr` method for a fresh correctly aligned ptr
+    // into memory for the target architecture.
+    next_free: usize,
+
+    target_alignment: u8,
+}
+
+impl Memory {
+    pub fn new(size: usize) -> Memory {
+        let alignment = align_of::<Vec>();
```
Comment (ltratt) [inline src/machine.rs]: Does this compile? Surely `Vec` requires a type paramater with it?

I'm also not sure that relying on an arbitrary struct (which is what `Vec` is) is safe. I think -- for every modern architecture -- you're safer with `align_of::<usize>`. [There were old machines where pointers and integers had different bit sizes. I'm not aware of any such machines being made in the last couple of decades or more, and I'm not sure anyone will make one again.]

File: src/machine.rs (hunk 8 above)
Comment (ltratt) [reply src/machine.rs:113]: Not freeing memory sounds like fragmentation to me!

File: src/machine.rs (hunk 10 above)
Comment (ltratt) [reply src/machine.rs]: One often needs two seemingly identical, but actually different, functions for these purposes. An example is here https://github.com/softdevteam/vob/pull/1/commits/d384a99d12a97c1e1824ca85353e3aaeb9df1b44#diff-b4aea3e418ccdb71239b96952d9cddb6R911

  * One function says "given an offset, tell me the aligned block that offset is *part of*".
  * The other function says "given an offset, tell me the first aligned block *after* that offset".

These two are easily confused IME.

File: src/machine.rs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    // not yet possible to free it.
+    pub fn next_aligned_ptr(&mut self, size: usize) -> usize {
+        if (self.next_free + size) >= self.memory.len() {
+            panic!("Out of memory.")
+        }
+
+        let r = self.next_free % self.target_alignment;
+        if r > 0 {
+            self.next_free + r
+        }
```
Comment (ltratt) [inline src/machine.rs]: The Rust way is to write `} else` on one line.

File: src/machine.rs [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        else {
+            self.next_free
+        }
+    }
+
+    fn write(&mut self, ptr: usize, value: &[u8]) {
```
Comment (ltratt) [inline src/machine.rs]: I'm not sure `write` is the clearest name. Or, at least, put a brief doc above the function to make clear what its purpose is? [I don't have a suggestions for a better name, I admit, but I was a bit confused when I first looked at this function, as it wasn't clear from the function paramaters what was being written to what.]

File: src/machine.rs [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        // Iterate over indexes, as we want to replace values, not shift them
+        for b in value {
+            self.memory[ptr] = b;
+            ptr += 1;
+        }
+        self.next_free = ptr
+    }
+
+    // TODO: It would be nice not to have to provide the size each time.
```
Comment (ltratt) [inline src/machine.rs:136]: How could you not provide the size?

File: src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,7 @@
...
+extern crate rustc;
+extern crate syntax;
+extern crate byteorder;
+
+pub mod interp;
```
Comment (ltratt) [inline src/lib.rs:6]: Is the interp file missing from this PR?

**Comment by ltratt** [comment]
I think there's a missing file (`interp.rs`) from this PR?

File: src/machine.rs (hunk 3 above)
Comment (jacob-hughes) [reply src/machine.rs:54]: Okay, this is something I shall ask about on the #rustc IRC.

File: src/machine.rs (hunk 5 above)
Comment (jacob-hughes) [reply src/machine.rs:77]: That's correct. I've added a comment for clarity.

File: src/machine.rs (hunk 11 above)
Comment (jacob-hughes) [reply src/machine.rs:61]: That's a good idea, I will probably do this in a future PR. I've added a TODO comment so that I don't forget this

File: src/lib.rs (hunk 1 above)
Comment (jacob-hughes) [reply src/lib.rs:1]: I have no issue adding the license in. Does this need to be in every file header? Or just the `lib.rs`?

File: src/machine.rs (hunk 14 above)
Comment (jacob-hughes) [reply src/machine.rs]: Ah okay! Fixed

File: src/machine.rs (hunk 2 above)
Comment (jacob-hughes) [reply src/machine.rs]: Fixed ;)

File: src/machine.rs (hunk 9 above)
Comment (jacob-hughes) [reply src/machine.rs]: Wow, dumb mistake. Thanks for the catch !

File: src/machine.rs (hunk 12 above)
Comment (jacob-hughes) [reply src/machine.rs]: Fixed

File: src/machine.rs (hunk 13 above)
Comment (jacob-hughes) [reply src/machine.rs]: That makes sense, I have changed this to `align_of::<usize>`

File: src/lib.rs (hunk 1 above)
Comment (ltratt) [reply src/lib.rs:1]: It needs to be in every file with meaningful content. [As a rough proxy, every `.rs` file.]

File: src/lib.rs (hunk 17 above)
Comment (jacob-hughes) [reply src/lib.rs:6]: It is yes, partly deliberately to keep the first PR focused on the machine layout, but in hindsight I should have at least provided the `struct` being referenced.

File: src/machine.rs (hunk 8 above)
Comment (vext01) [reply src/machine.rs:113]: If you don't free your memory, it cannot fragment, but you do have a leaking problem!

File: src/machine.rs (hunk 10 above)
Comment (vext01) [reply src/machine.rs]: Jake's intention is the latter.

**Comment by vext01** [comment]
I've now reviewed the whole change. I'll review the missing file once it appears.

File: src/machine.rs (hunk 15 above)
Comment (jacob-hughes) [reply src/machine.rs]: I've changed this to `store`. In my mind this might be slightly clearer, what do you think?

File: src/machine.rs (hunk 16 above)
Comment (vext01) [reply src/machine.rs:136]: `value.len()`, maybe?

https://doc.rust-lang.org/std/primitive.slice.html#method.len

File: src/machine.rs (hunk 15 above)
Comment (ltratt) [reply src/machine.rs]: Works for me.

**Comment by jacob-hughes** [comment]
Okay, think I'm ready for a re-review

File: src/machine.rs (hunk 8 above)
Comment (ltratt) [reply src/machine.rs:113]: Point taken... though it's very much tomato and tomato ;)

File: .travis.yml [hunk 18]
```diff
@@ -0,0 +1,7 @@
+language: rust
+rust: nightly
+cache: cargo
+script:
```
Comment (ltratt) [inline .travis.yml]: I think travis does `cargo test` by default which, I think, is what you want.

File: .travis.yml [hunk 19]
```diff
@@ -0,0 +1,7 @@
+language: rust
+rust: nightly
+cache: cargo
```
Comment (ltratt) [inline .travis.yml]: I think this is the default.

File: src/machine.rs (hunk 4 above)
Comment (vext01) [reply src/machine.rs:70]: [In general]

File: src/interp.rs [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use rustc::ty::Ty;
+
+pub struct TypedVal<'tcx> {
```
Comment (vext01) [inline src/interp.rs]: What's this for? Comment it up?

File: src/machine.rs (hunk 3 above)
Comment (vext01) [reply src/machine.rs:54]: Raise an issue so it's not forgotten?

File: src/machine.rs (hunk 11 above)
Comment (vext01) [reply src/machine.rs:61]: Again, if we raise an issue for all of the things that we want to implement in later PRs, that would be great.

Comment (jacob-hughes) [reply src/machine.rs:61]: That sounds sensible.

File: src/interp.rs (hunk 20 above)
Comment (jacob-hughes) [reply src/interp.rs]: Fixed

**Comment by jacob-hughes** [comment]
Think I've fixed all the issues mentioned. Ready to squash before merge if you are both happy?

File: src/machine.rs (hunk 16 above)
Comment (vext01) [reply src/machine.rs:136]: So can you just use the length of the slice?

File: src/machine.rs [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+    fn read(&self, ptr: usize, size: usize) -> &[u8] {
+        let end = ptr + size;
+        &self.memory[ptr..end]
+    }
+}
+
+pub struct Machine<'tcx> {
```
Comment (vext01) [inline src/machine.rs]: Let's add a brief comment to this struct.

**Comment by vext01** [comment]
A couple of small comments, then we are good to go.

File: src/machine.rs (hunk 16 above)
Comment (jacob-hughes) [reply src/machine.rs:136]: @vext01 I'm not quite sure what you're referring to here?

The `read` method will read `size` bytes from a given offset in the memory Vec. What slice would I use the length of?

Comment (vext01) [reply src/machine.rs:136]: Sorry, I misread.

So now, back to Laurie's question. How could you read without knowing how many bytes to read?

**Comment by jacob-hughes** [comment]
Fixed :) Should be good to go.

**Comment by vext01** [comment]
If @ltratt is ready, I think you can squash. Laurie?

File: .travis.yml [hunk 22]
```diff
@@ -2,6 +2,5 @@ language: rust
 rust: nightly
 cache: cargo
 script:
-    - cargo build
-    - cargo check
+    - cargo test
```
Comment (ltratt) [inline .travis.yml]: I'm not convinced you even need this. I have travis files which just say `rust: nightly` or `rust: stable` and everything works.

File: src/machine.rs [hunk 23]
```diff
@@ -140,6 +140,7 @@ impl Memory {
     }
 }
 
+// The VM which stores the state of interpreting the MIR
```
Comment (ltratt) [inline src/machine.rs]: This is pretty hard to parse (and should it be `///`?). Can we try rephrasing?

**Comment by ltratt** [comment]
I have two minor comments -- nothing that will take much time to address.

File: .travis.yml (hunk 22 above)
Comment (jacob-hughes) [reply .travis.yml]: Ah I misunderstood, I'll give it a go without.

File: src/machine.rs (hunk 23 above)
Comment (jacob-hughes) [reply src/machine.rs]: I've given it another go, it's a little hard to come up with a comment for this without saying anything that sounds 'too obvious'.

**Comment by ltratt** [comment]
Works for me! Time to squash.

**Comment by jacob-hughes** [comment]
Squashed
